File: code/ex286.java
class Solution {
    public final int[][] dir = new int[][]{{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
    
    public void wallsAndGates(int[][] rooms) {
        int m = rooms.length;    if (m == 0)    return;
        int n = rooms[0].length;    if (n == 0)    return;
        
        Queue<int[]> queue = new LinkedList<>();
        for (int i = 0; i < m; i++)
            for (int j = 0; j < n; j++)
                if (rooms[i][j] == 0)
                    queue.offer(new int[]{i, j});
        
        int dis = 0;
        while (!queue.isEmpty()) {
            for (int z = queue.size(); z > 0; z--) {
                int[] out = queue.poll();
                int i = out[0], j = out[1];
                for (int[] d : dir) {
                    int ni = i + d[0], nj = j + d[1];
                    if (ni < 0 || nj < 0 || ni >= m || nj >= n || rooms[ni][nj] != Integer.MAX_VALUE)
                        continue;
                    rooms[ni][nj] = dis + 1;
                    queue.offer(new int[]{ni, nj});
                }
            }
            dis++;
        }
    }
}   

File: code/ex286.cpp
class Solution {
public:
    void wallsAndGates(vector<vector<int>>& rooms) {
        const int dir[4][2] = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
        const int m = rooms.size();    if (m == 0)    return;
        const int n = rooms[0].size();    if (n == 0)    return;
        
        queue<vector<int>> que;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (rooms[i][j] == 0) {
                    que.push({i, j});
                }
            }
        }
        
        int step = 1;
        while (!que.empty()) {
            int size = que.size();
            for (int z = size; z > 0; z--) {
                vector<int> out = que.front();   que.pop();
                int i = out[0], j = out[1];
                for (auto& d : dir) {
                    int ni = i + d[0], nj = j + d[1];
                    if (ni < 0 || nj < 0 || ni >= m || nj >= n || rooms[ni][nj] != INT_MAX)
                        continue;
                    rooms[ni][nj] = step;
                    que.push({ni, nj});
                }
            }
            step++;
        }
    }
};